File: backend/backup_20260108/utils/duplicate_checker.py
#!/usr/bin/env python3
"""
중복 검사를 위한 공통 유틸리티 함수
모든 데이터 처리 스크립트에서 재사용 가능
"""

import os
import json
from typing import Set, List, Dict, Any, Callable


def load_processed_urls(file_path: str) -> Set[str]:
    """
    JSONL 파일에서 처리된 youtube_link 추출
    
    Args:
        file_path: JSONL 파일 경로
    
    Returns:
        처리된 URL들의 set
    
    Example:
        >>> processed_urls = load_processed_urls('tzuyang_restaurant_results.jsonl')
        >>> print(f"처리된 URL: {len(processed_urls)}개")
    """
    urls = set()
    
    if not os.path.exists(file_path):
        return urls
    
    with open(file_path, 'r', encoding='utf-8') as f:
        for line_num, line in enumerate(f, 1):
            line = line.strip()
            if not line:
                continue
            
            try:
                data = json.loads(line)
                if 'youtube_link' in data:
                    urls.add(data['youtube_link'])
            except json.JSONDecodeError as e:
                print(f"⚠️  JSON 파싱 오류 (라인 {line_num}): {e}")
                continue
    
    return urls


def load_processed_restaurants(
    file_path: str, 
    key: str = 'name',
    nested_key: str = 'restaurants'
) -> Set[str]:
    """
    JSONL 파일에서 처리된 restaurant 이름/ID 추출
    
    Args:
        file_path: JSONL 파일 경로
        key: 추출할 키 ('name', 'unique_id' 등)
        nested_key: restaurants가 들어있는 상위 키 (기본값: 'restaurants')
    
    Returns:
        처리된 restaurant 키값들의 set
    
    Example:
        >>> processed = load_processed_restaurants('output.jsonl', key='name')
        >>> new_restaurants = [r for r in input_list if r['name'] not in processed]
    """
    restaurants = set()
    
    if not os.path.exists(file_path):
        return restaurants
    
    with open(file_path, 'r', encoding='utf-8') as f:
        for line_num, line in enumerate(f, 1):
            line = line.strip()
            if not line:
                continue
            
            try:
                data = json.loads(line)
                
                # nested_key가 있는 경우 (예: {'restaurants': [...]})
                if nested_key and nested_key in data:
                    for restaurant in data[nested_key]:
                        if key in restaurant and restaurant[key]:
                            restaurants.add(restaurant[key])
                # nested_key가 없는 경우 (예: 각 라인이 restaurant)
                elif not nested_key and key in data and data[key]:
                    restaurants.add(data[key])
                    
            except json.JSONDecodeError as e:
                print(f"⚠️  JSON 파싱 오류 (라인 {line_num}): {e}")
                continue
    
    return restaurants


def load_processed_unique_ids(file_path: str) -> Set[str]:
    """
    JSONL 파일에서 처리된 unique_id 추출
    
    Args:
        file_path: JSONL 파일 경로
    
    Returns:
        처리된 unique_id들의 set
    
    Example:
        >>> written_ids = load_processed_unique_ids('transforms.jsonl')
        >>> if unique_id not in written_ids:
        >>>     # 새 데이터 추가
    """
    ids = set()
    
    if not os.path.exists(file_path):
        return ids
    
    with open(file_path, 'r', encoding='utf-8') as f:
        for line_num, line in enumerate(f, 1):
            line = line.strip()
            if not line:
                continue
            
            try:
                data = json.loads(line)
                if 'unique_id' in data and data['unique_id']:
                    ids.add(data['unique_id'])
            except json.JSONDecodeError as e:
                print(f"⚠️  JSON 파싱 오류 (라인 {line_num}): {e}")
                continue
    
    return ids


def filter_new_items(
    input_items: List[Any], 
    processed_items: Set[str], 
    key_func: Callable[[Any], str]
) -> List[Any]:
    """
    중복되지 않은 항목만 필터링
    
    Args:
        input_items: 입력 항목 리스트
        processed_items: 이미 처리된 항목 set
        key_func: 항목에서 키를 추출하는 함수
    
    Returns:
        중복되지 않은 항목 리스트
    
    Example:
        >>> processed_urls = load_processed_urls('output.jsonl')
        >>> new_items = filter_new_items(
        >>>     input_list, 
        >>>     processed_urls, 
        >>>     lambda x: x['youtube_link']
        >>> )
    """
    return [
        item for item in input_items 
        if key_func(item) not in processed_items
    ]


def append_to_jsonl(
    file_path: str, 
    data: Any,
    create_dirs: bool = True
) -> int:
    """
    JSONL 파일에 데이터 추가 (append 모드)
    
    Args:
        file_path: JSONL 파일 경로
        data: 추가할 데이터 (dict 또는 list)
        create_dirs: 디렉토리가 없으면 생성할지 여부
    
    Returns:
        추가된 항목 수
    
    Example:
        >>> count = append_to_jsonl('output.jsonl', new_data)
        >>> print(f"{count}개 항목 추가됨")
    """
    count = 0
    
    # 디렉토리가 없으면 생성
    if create_dirs:
        dir_path = os.path.dirname(file_path)
        if dir_path and not os.path.exists(dir_path):
            os.makedirs(dir_path, exist_ok=True)
    
    with open(file_path, 'a', encoding='utf-8') as f:
        if isinstance(data, list):
            for item in data:
                f.write(json.dumps(item, ensure_ascii=False) + '\n')
                count += 1
        else:
            f.write(json.dumps(data, ensure_ascii=False) + '\n')
            count = 1
    
    return count


def load_multiple_processed_urls(*file_paths: str) -> Set[str]:
    """
    여러 JSONL 파일에서 처리된 youtube_link 추출
    
    Args:
        *file_paths: JSONL 파일 경로들 (가변 인자)
    
    Returns:
        모든 파일에서 추출한 URL들의 합집합
    
    Example:
        >>> all_processed = load_multiple_processed_urls(
        >>>     'results.jsonl',
        >>>     'errors.jsonl',
        >>>     'no_selection.jsonl'
        >>> )
    """
    all_urls = set()
    
    for file_path in file_paths:
        urls = load_processed_urls(file_path)
        all_urls.update(urls)
    
    return all_urls


def load_multiple_processed_restaurants(
    *file_paths: str,
    key: str = 'name',
    nested_key: str = 'restaurants'
) -> Set[str]:
    """
    여러 JSONL 파일에서 처리된 restaurant 이름/ID 추출
    
    Args:
        *file_paths: JSONL 파일 경로들 (가변 인자)
        key: 추출할 키 ('name', 'unique_id' 등)
        nested_key: restaurants가 들어있는 상위 키
    
    Returns:
        모든 파일에서 추출한 restaurant 키값들의 합집합
    
    Example:
        >>> all_processed = load_multiple_processed_restaurants(
        >>>     'results.jsonl',
        >>>     'errors.jsonl',
        >>>     key='name'
        >>> )
    """
    all_restaurants = set()
    
    for file_path in file_paths:
        restaurants = load_processed_restaurants(file_path, key, nested_key)
        all_restaurants.update(restaurants)
    
    return all_restaurants


def get_file_stats(file_path: str) -> Dict[str, Any]:
    """
    JSONL 파일의 통계 정보 반환
    
    Args:
        file_path: JSONL 파일 경로
    
    Returns:
        파일 통계 정보 딕셔너리
    
    Example:
        >>> stats = get_file_stats('output.jsonl')
        >>> print(f"총 {stats['total_lines']}개 항목, 파일 크기: {stats['size_mb']:.2f}MB")
    """
    if not os.path.exists(file_path):
        return {
            'exists': False,
            'total_lines': 0,
            'valid_lines': 0,
            'invalid_lines': 0,
            'size_bytes': 0,
            'size_mb': 0
        }
    
    total_lines = 0
    valid_lines = 0
    invalid_lines = 0
    
    with open(file_path, 'r', encoding='utf-8') as f:
        for line in f:
            if line.strip():
                total_lines += 1
                try:
                    json.loads(line)
                    valid_lines += 1
                except json.JSONDecodeError:
                    invalid_lines += 1
    
    file_size = os.path.getsize(file_path)
    
    return {
        'exists': True,
        'total_lines': total_lines,
        'valid_lines': valid_lines,
        'invalid_lines': invalid_lines,
        'size_bytes': file_size,
        'size_mb': file_size / (1024 * 1024)
    }


if __name__ == '__main__':
    # 테스트 코드
    print("🧪 중복 검사 유틸리티 함수 테스트\n")
    
    # 예시 파일 경로
    test_file = 'test_output.jsonl'
    
    # 테스트 데이터 작성
    test_data = [
        {'youtube_link': 'https://youtube.com/watch?v=1', 'unique_id': 'id1', 'name': 'Restaurant 1'},
        {'youtube_link': 'https://youtube.com/watch?v=2', 'unique_id': 'id2', 'name': 'Restaurant 2'},
    ]
    
    print(f"📝 테스트 파일 작성: {test_file}")
    append_to_jsonl(test_file, test_data, create_dirs=False)
    
    # 통계 확인
    stats = get_file_stats(test_file)
    print(f"\n📊 파일 통계:")
    print(f"   - 총 라인: {stats['total_lines']}")
    print(f"   - 유효 라인: {stats['valid_lines']}")
    print(f"   - 파일 크기: {stats['size_mb']:.4f}MB")
    
    # URL 로드
    urls = load_processed_urls(test_file)
    print(f"\n🔗 처리된 URL: {len(urls)}개")
    for url in urls:
        print(f"   - {url}")
    
    # unique_id 로드
    ids = load_processed_unique_ids(test_file)
    print(f"\n🆔 처리된 unique_id: {len(ids)}개")
    for uid in ids:
        print(f"   - {uid}")
    
    # 테스트 파일 삭제
    if os.path.exists(test_file):
        os.remove(test_file)
        print(f"\n🗑️  테스트 파일 삭제: {test_file}")
    
    print("\n✅ 테스트 완료!")

File: backend/backup-20260108/utils/duplicate-checker.ts
/**
 * 중복 검사를 위한 공통 유틸리티 함수
 * 모든 데이터 처리 스크립트에서 재사용 가능
 */

import fs from 'fs';
import path from 'path';

/**
 * JSONL 파일에서 처리된 youtube_link 추출
 * 
 * @param filePath JSONL 파일 경로
 * @returns 처리된 URL들의 Set
 * 
 * @example
 * const processedUrls = loadProcessedUrls('tzuyang_restaurant_results.jsonl');
 * console.log(`처리된 URL: ${processedUrls.size}개`);
 */
export function loadProcessedUrls(filePath: string): Set<string> {
  const urls = new Set<string>();
  
  if (!fs.existsSync(filePath)) {
    return urls;
  }
  
  try {
    const content = fs.readFileSync(filePath, 'utf8');
    const lines = content.split('\n');
    
    for (let lineNum = 0; lineNum < lines.length; lineNum++) {
      const line = lines[lineNum].trim();
      if (!line) continue;
      
      try {
        const data = JSON.parse(line);
        if (data.youtube_link) {
          urls.add(data.youtube_link);
        }
      } catch (err) {
        console.warn(`⚠️  JSON 파싱 오류 (라인 ${lineNum + 1}):`, err instanceof Error ? err.message : String(err));
      }
    }
  } catch (err) {
    console.error(`❌ 파일 읽기 실패 (${filePath}):`, err instanceof Error ? err.message : String(err));
  }
  
  return urls;
}

/**
 * JSONL 파일에서 처리된 restaurant 이름/ID 추출
 * 
 * @param filePath JSONL 파일 경로
 * @param key 추출할 키 ('name', 'unique_id' 등)
 * @param nestedKey restaurants가 들어있는 상위 키 (기본값: 'restaurants')
 * @returns 처리된 restaurant 키값들의 Set
 * 
 * @example
 * const processed = loadProcessedRestaurants('output.jsonl', 'name');
 * const newRestaurants = inputList.filter(r => !processed.has(r.name));
 */
export function loadProcessedRestaurants(
  filePath: string,
  key: string = 'name',
  nestedKey: string | null = 'restaurants'
): Set<string> {
  const restaurants = new Set<string>();
  
  if (!fs.existsSync(filePath)) {
    return restaurants;
  }
  
  try {
    const content = fs.readFileSync(filePath, 'utf8');
    const lines = content.split('\n');
    
    for (let lineNum = 0; lineNum < lines.length; lineNum++) {
      const line = lines[lineNum].trim();
      if (!line) continue;
      
      try {
        const data = JSON.parse(line);
        
        // nestedKey가 있는 경우 (예: {'restaurants': [...]})
        if (nestedKey && data[nestedKey]) {
          const items = data[nestedKey];
          if (Array.isArray(items)) {
            for (const restaurant of items) {
              if (restaurant[key]) {
                restaurants.add(restaurant[key]);
              }
            }
          }
        }
        // nestedKey가 없는 경우 (예: 각 라인이 restaurant)
        else if (!nestedKey && data[key]) {
          restaurants.add(data[key]);
        }
      } catch (err) {
        console.warn(`⚠️  JSON 파싱 오류 (라인 ${lineNum + 1}):`, err instanceof Error ? err.message : String(err));
      }
    }
  } catch (err) {
    console.error(`❌ 파일 읽기 실패 (${filePath}):`, err instanceof Error ? err.message : String(err));
  }
  
  return restaurants;
}

/**
 * JSONL 파일에서 처리된 unique_id 추출
 * 
 * @param filePath JSONL 파일 경로
 * @returns 처리된 unique_id들의 Set
 * 
 * @example
 * const writtenIds = loadProcessedUniqueIds('transforms.jsonl');
 * if (!writtenIds.has(uniqueId)) {
 *   // 새 데이터 추가
 * }
 */
export function loadProcessedUniqueIds(filePath: string): Set<string> {
  const ids = new Set<string>();
  
  if (!fs.existsSync(filePath)) {
    return ids;
  }
  
  try {
    const content = fs.readFileSync(filePath, 'utf8');
    const lines = content.split('\n');
    
    for (let lineNum = 0; lineNum < lines.length; lineNum++) {
      const line = lines[lineNum].trim();
      if (!line) continue;
      
      try {
        const data = JSON.parse(line);
        if (data.unique_id) {
          ids.add(data.unique_id);
        }
      } catch (err) {
        console.warn(`⚠️  JSON 파싱 오류 (라인 ${lineNum + 1}):`, err instanceof Error ? err.message : String(err));
      }
    }
  } catch (err) {
    console.error(`❌ 파일 읽기 실패 (${filePath}):`, err instanceof Error ? err.message : String(err));
  }
  
  return ids;
}

/**
 * 중복되지 않은 항목만 필터링
 * 
 * @param inputItems 입력 항목 리스트
 * @param processedItems 이미 처리된 항목 Set
 * @param keyFunc 항목에서 키를 추출하는 함수
 * @returns 중복되지 않은 항목 리스트
 * 
 * @example
 * const processedUrls = loadProcessedUrls('output.jsonl');
 * const newItems = filterNewItems(
 *   inputList,
 *   processedUrls,
 *   (x) => x.youtube_link
 * );
 */
export function filterNewItems<T>(
  inputItems: T[],
  processedItems: Set<string>,
  keyFunc: (item: T) => string
): T[] {
  return inputItems.filter(item => !processedItems.has(keyFunc(item)));
}

/**
 * JSONL 파일에 데이터 추가 (append 모드)
 * 
 * @param filePath JSONL 파일 경로
 * @param data 추가할 데이터 (객체 또는 배열)
 * @param createDirs 디렉토리가 없으면 생성할지 여부 (기본값: true)
 * @returns 추가된 항목 수
 * 
 * @example
 * const count = appendToJsonl('output.jsonl', newData);
 * console.log(`${count}개 항목 추가됨`);
 */
export function appendToJsonl(
  filePath: string,
  data: any | any[],
  createDirs: boolean = true
): number {
  let count = 0;
  
  // 디렉토리가 없으면 생성
  if (createDirs) {
    const dirPath = path.dirname(filePath);
    if (dirPath && !fs.existsSync(dirPath)) {
      fs.mkdirSync(dirPath, { recursive: true });
    }
  }
  
  const items = Array.isArray(data) ? data : [data];
  
  try {
    const stream = fs.createWriteStream(filePath, { flags: 'a', encoding: 'utf8' });
    
    for (const item of items) {
      stream.write(JSON.stringify(item) + '\n');
      count++;
    }
    
    stream.end();
  } catch (err) {
    console.error(`❌ 파일 쓰기 실패 (${filePath}):`, err instanceof Error ? err.message : String(err));
    throw err;
  }
  
  return count;
}

/**
 * 여러 JSONL 파일에서 처리된 youtube_link 추출
 * 
 * @param filePaths JSONL 파일 경로들
 * @returns 모든 파일에서 추출한 URL들의 합집합
 * 
 * @example
 * const allProcessed = loadMultipleProcessedUrls(
 *   'results.jsonl',
 *   'errors.jsonl',
 *   'no_selection.jsonl'
 * );
 */
export function loadMultipleProcessedUrls(...filePaths: string[]): Set<string> {
  const allUrls = new Set<string>();
  
  for (const filePath of filePaths) {
    const urls = loadProcessedUrls(filePath);
    urls.forEach(url => allUrls.add(url));
  }
  
  return allUrls;
}

/**
 * 여러 JSONL 파일에서 처리된 restaurant 이름/ID 추출
 * 
 * @param filePaths JSONL 파일 경로들
 * @param key 추출할 키 ('name', 'unique_id' 등)
 * @param nestedKey restaurants가 들어있는 상위 키
 * @returns 모든 파일에서 추출한 restaurant 키값들의 합집합
 * 
 * @example
 * const allProcessed = loadMultipleProcessedRestaurants(
 *   ['results.jsonl', 'errors.jsonl'],
 *   'name'
 * );
 */
export function loadMultipleProcessedRestaurants(
  filePaths: string[],
  key: string = 'name',
  nestedKey: string | null = 'restaurants'
): Set<string> {
  const allRestaurants = new Set<string>();
  
  for (const filePath of filePaths) {
    const restaurants = loadProcessedRestaurants(filePath, key, nestedKey);
    restaurants.forEach(restaurant => allRestaurants.add(restaurant));
  }
  
  return allRestaurants;
}

/**
 * JSONL 파일의 통계 정보 반환
 * 
 * @param filePath JSONL 파일 경로
 * @returns 파일 통계 정보
 * 
 * @example
 * const stats = getFileStats('output.jsonl');
 * console.log(`총 ${stats.totalLines}개 항목, 파일 크기: ${stats.sizeMB.toFixed(2)}MB`);
 */
export function getFileStats(filePath: string): {
  exists: boolean;
  totalLines: number;
  validLines: number;
  invalidLines: number;
  sizeBytes: number;
  sizeMB: number;
} {
  if (!fs.existsSync(filePath)) {
    return {
      exists: false,
      totalLines: 0,
      validLines: 0,
      invalidLines: 0,
      sizeBytes: 0,
      sizeMB: 0
    };
  }
  
  let totalLines = 0;
  let validLines = 0;
  let invalidLines = 0;
  
  try {
    const content = fs.readFileSync(filePath, 'utf8');
    const lines = content.split('\n');
    
    for (const line of lines) {
      if (line.trim()) {
        totalLines++;
        try {
          JSON.parse(line);
          validLines++;
        } catch {
          invalidLines++;
        }
      }
    }
  } catch (err) {
    console.error(`❌ 파일 읽기 실패 (${filePath}):`, err instanceof Error ? err.message : String(err));
  }
  
  const fileSize = fs.statSync(filePath).size;
  
  return {
    exists: true,
    totalLines,
    validLines,
    invalidLines,
    sizeBytes: fileSize,
    sizeMB: fileSize / (1024 * 1024)
  };
}

// 테스트 코드 (직접 실행 시)
if (import.meta.url === `file://${process.argv[1]}`) {
  console.log('🧪 중복 검사 유틸리티 함수 테스트\n');
  
  const testFile = 'test_output.jsonl';
  
  // 테스트 데이터 작성
  const testData = [
    { youtube_link: 'https://youtube.com/watch?v=1', unique_id: 'id1', name: 'Restaurant 1' },
    { youtube_link: 'https://youtube.com/watch?v=2', unique_id: 'id2', name: 'Restaurant 2' },
  ];
  
  console.log(`📝 테스트 파일 작성: ${testFile}`);
  appendToJsonl(testFile, testData, false);
  
  // 통계 확인
  const stats = getFileStats(testFile);
  console.log('\n📊 파일 통계:');
  console.log(`   - 총 라인: ${stats.totalLines}`);
  console.log(`   - 유효 라인: ${stats.validLines}`);
  console.log(`   - 파일 크기: ${stats.sizeMB.toFixed(4)}MB`);
  
  // URL 로드
  const urls = loadProcessedUrls(testFile);
  console.log(`\n🔗 처리된 URL: ${urls.size}개`);
  urls.forEach(url => console.log(`   - ${url}`));
  
  // unique_id 로드
  const ids = loadProcessedUniqueIds(testFile);
  console.log(`\n🆔 처리된 unique_id: ${ids.size}개`);
  ids.forEach(id => console.log(`   - ${id}`));
  
  // 테스트 파일 삭제
  if (fs.existsSync(testFile)) {
    fs.unlinkSync(testFile);
    console.log(`\n🗑️  테스트 파일 삭제: ${testFile}`);
  }
  
  console.log('\n✅ 테스트 완료!');
}
